File: aetherment/plugin/Gui/Window/Aetherment/Explorer/Tree.cs
using System;
using System.Collections.Generic;
using System.Linq;
using ImGuiNET;

namespace Aetherment.Gui.Window.Aetherment.Explorer;

public class Tree {
	private class Node {
		public bool isFolder;
		public bool enabled;
		public Node? parent;
		public string? path;
		public SortedDictionary<string, Node> children;
		
		public Node(Node? parent = null, string? path = null, bool isFolder = true) {
			this.isFolder = isFolder;
			enabled = true;
			this.parent = parent;
			this.path = path;
			children = new();
		}
	}
	
	public string SelectedPath;
	private string name;
	private Node nodes;
	private Action<string> callback;
	
	public Tree(string name, Action<string> callback) {
		SelectedPath = "";
		this.name = name;
		nodes = new();
		this.callback = callback;
	}
	
	public void AddNode(string path) {
		lock(nodes) {
			var segs = path.Split('/');
			var curnode = nodes;
			for(int i = 0; i < segs.Length - 1; i++) {
				var seg = segs[i];
				if(!curnode.children!.TryGetValue(seg, out var node)) {
					node = new(curnode);
					curnode.children.Add(seg, node);
				}
				
				curnode = node;
			}
			curnode.children!.Add(segs.Last(), new(curnode, path, false));
		}
	}
	
	public void SetNodeState(bool state) {
		void setstate(Node node) {
			node.enabled = state;
			foreach(var n in node.children.Values)
				setstate(n);
		}
		
		foreach(var n in nodes.children.Values)
			setstate(n);
	}
	
	public void SetNodeState(string path, bool state) {
		var node = nodes;
		foreach(var seg in path.Split('/'))
			node = node.children[seg];
		SetNodeState(node, state);
	}
	
	private void SetNodeState(Node node, bool state) {
		if(node == nodes || node.enabled == state)
			return;
		
		node.enabled = state;
		if(node.parent!.children.Any(n => n.Value.enabled) != node.parent!.enabled)
			SetNodeState(node.parent, state);
	}
	
	public void Draw() {
		lock(nodes)
			DrawNode(name, nodes);
	}
	
	private void DrawNode(string name, Node node) {
		if(!node.isFolder) {
			if(!node.enabled)
				ImGui.PushStyleColor(ImGuiCol.Text, ImGui.GetColorU32(ImGuiCol.TextDisabled));
			if(ImGui.Selectable(name, SelectedPath == node.path)) {
				SelectedPath = node.path!;
				callback(SelectedPath);
			}
			if(!node.enabled)
				ImGui.PopStyleColor();
			return;
		}
		
		if(!node.enabled)
			ImGui.PushStyleColor(ImGuiCol.Text, ImGui.GetColorU32(ImGuiCol.TextDisabled));
		var treeOpen = ImGui.TreeNode(name);
		if(!node.enabled)
				ImGui.PopStyleColor();
		
		if(treeOpen) {
			foreach(var c in node.children)
				if(!c.Value.isFolder)
					DrawNode(c.Key, c.Value);
			foreach(var c in node.children)
				if(c.Value.isFolder)
					DrawNode(c.Key, c.Value);
			ImGui.TreePop();
		}
	}
}

File: aetherment/src/gui/window/aetherment/explorer/tree.rs
use std::{collections::BTreeMap, rc::Rc};
use crate::gui::aeth;

pub struct Tree {
	selected: usize,
	nodes: Vec<Node>,
}

impl Tree {
	pub fn new<S>(name: S) -> Self where
	S: Into<String> {
		Tree {
			selected: usize::MAX,
			nodes: vec![Node {
				name: name.into(),
				is_folder: true,
				is_enabled: true,
				parent: None,
				children: BTreeMap::new(),
			}],
		}
	}
	
	pub fn add_node(&mut self, path: &str) {
		let mut curnode = 0;
		for seg in path.split('/') {
			curnode = if !self.nodes[curnode].children.contains_key(seg) {
				let i = self.nodes.len();
				let node = Node {
					name: seg.to_owned(),
					is_folder: true,
					is_enabled: true,
					parent: Some(curnode),
					children: BTreeMap::new(),
				};
				self.nodes[curnode].children.insert(Rc::from(node.name.as_str()), i);
				self.nodes.push(node);
				i
			} else {
				*self.nodes[curnode].children.get(seg).unwrap()
			}
		}
		self.nodes[curnode].is_folder = false;
	}
	
	pub fn node_state(&mut self, path: &str, state: bool) {
		let mut curnode = 0;
		for seg in path.split('/') {
			curnode = *self.nodes[curnode].children.get(seg).unwrap();
		}
		self.node_state_i(curnode, state);
	}
	
	fn node_state_i(&mut self, node: usize, state: bool) {
		let mut node = self.nodes.get_mut(node).unwrap();
		if node.is_enabled != state && node.parent != None {
			node.is_enabled = state;
			
			let pi = node.parent.unwrap();
			let p = self.nodes.get(pi).unwrap();
			if p.children.values().any(|n| self.nodes[*n].is_enabled != p.is_enabled) {
				self.node_state_i(pi, state);
			}
		}
	}
	
	pub fn node_state_all(&mut self, state: bool) {
		let mut iter = self.nodes.iter_mut();
		iter.next(); // skip root node
		iter.for_each(|n| n.is_enabled = state);
	}
	
	pub fn draw(&mut self) -> Option<(imgui::MouseButton, String)> {
		let mut r = None;
		if let Some((button, id)) = self.draw_node(0) {
			r = Some((button, self.node_path(id)));
			self.selected = id;
		}
		r
	}
	
	fn draw_node(&self, nodeid: usize) -> Option<(imgui::MouseButton, usize)> {
		let mut r = None;
		let node = self.nodes.get(nodeid).unwrap();
		if node.is_folder {
			if !node.is_enabled {imgui::push_style_color(imgui::Col::Text, imgui::get_color(imgui::Col::TextDisabled))}
			if !aeth::tree(&node.name, || {
				if !node.is_enabled {imgui::pop_style_color(1)}
				node.children.iter().for_each(|(_, n)| if self.nodes[*n].is_folder {r = r.or(self.draw_node(*n))});
				node.children.iter().for_each(|(_, n)| if !self.nodes[*n].is_folder {r = r.or(self.draw_node(*n))});
			}) && !node.is_enabled {imgui::pop_style_color(1)}
		} else {
			if imgui::selectable(
				&node.name,
				nodeid == self.selected,
				if node.is_enabled {imgui::SelectableFlags::None} else {imgui::SelectableFlags::Disabled},
				[0.0, 0.0]
			) {
				r = Some((imgui::MouseButton::Left, nodeid))
			} else if imgui::is_item_clicked(imgui::MouseButton::Right) {
				r = Some((imgui::MouseButton::Right, nodeid))
			}
		}
		
		// if imgui::is_item_hovered() {
		// 	aeth::tooltip(|| {
		// 		imgui::text(&self.node_path(nodeid));
		// 	});
		// }
		
		r
	}
	
	fn node_path(&self, node: usize) -> String {
		let mut node = self.nodes.get(node).unwrap();
		let mut path = node.name.to_owned();
		while let Some(p) = node.parent && p != 0 {
			node = self.nodes.get(p).unwrap();
			path.insert(0, '/');
			path.insert_str(0, &node.name);
		}
		path
	}
}

pub struct Node {
	name: String,
	is_folder: bool,
	is_enabled: bool,
	parent: Option<usize>,
	children: BTreeMap<Rc<str>, usize>,
}